srml/support/src/double_map.rs

```diff
@@ -38,42 +39,78 @@ pub trait StorageDoubleMap {
 	const PREFIX: &'static [u8];
 
 	/// Insert an entry into this map.
-	fn insert(k1: Self::Key1, k2: Self::Key2, val: Self::Value) {
+	fn insert<Q, R>(k1: &Q, k2: &R, val: Self::Value)
+	where
+		Self::Key1: Borrow<Q>,
+		Self::Key2: Borrow<R>,
+		Q: Codec,
+		R: Codec
+	{
 		unhashed::put(&Self::full_key(k1, k2)[..], &val);
 	}
 
 	/// Remove an entry from this map.
-	fn remove(k1: Self::Key1, k2: Self::Key2) {
+	fn remove<Q, R>(k1: &Q, k2: &R)
+	where
+		Self::Key1: Borrow<Q>,
+		Self::Key2: Borrow<R>,
+		Q: Codec,
```

Comment:
why do `Q` and `R` need to be `Codec` if `Key1`/`Key2` can be borrowed from them anyway (and are already `Codec`)?

Reply:
It tells the compiler that the borrowed key types (`Q` and `R`) implement Codec as well.
I followed the pattern used by `std::collections::HashMap` .e.g
```
    pub fn get<Q: ?Sized>(&self, k: &Q) -> Option<&V>
        where K: Borrow<Q>,
              Q: Hash + Eq
```
https://github.com/rust-lang/rust/blob/b244f61b77c42d7be695afd7901ee4418559e518/src/libstd/collections/hash/map.rs#L1300-L1302

Reply:
Guess it's just something that `full_key` forces you into.